File: calculator/src/main/java/com/turlir/calculator/extractors/ExpressionPartExtractor.java
package com.turlir.calculator.extractors;

import java.util.Iterator;

/**
 * Простое выделение из математического выражения операторов и операндов <br>
 * Выполняет чистку строкового представления выражения
 */
public class ExpressionPartExtractor implements IntervalExtractor {

    @Override
    public Iterator<Interval> iterator(String value) {
        return new IntervalIterator(value);
    }

    private static class IntervalIterator implements Iterator<Interval> {

        private final String mStr;
        private int mLength;

        private IntervalIterator(String str) {
            mStr = str.replaceAll("\\s+", "").replaceAll(",", ".");
        }

        @Override
        public Interval next() {
            Interval part = new Interval(mStr, mLength);
            mLength += part.value.length();
            return part;
        }

        @Override
        public boolean hasNext() {
            return mLength < mStr.length();
        }

    }
}

File: calculator/src/main/java/com/turlir/calculator/extractors/ExpressionPartExtractor.kt
package com.turlir.calculator.extractors

/**
 * Простое выделение из математического выражения операторов и операндов <br></br>
 * Выполняет чистку строкового представления выражения
 */
class ExpressionPartExtractor : IntervalExtractor {

    override fun iterator(value: String): Iterator<Interval> {
        return IntervalIterator(value)
    }

    private class IntervalIterator constructor(str: String) : Iterator<Interval> {

        companion object {
            private val oneRegex = "\\s+".toRegex()
            private val twoRegex = ",".toRegex()
        }

        private val mStr: String = str.replace(oneRegex, "").replace(twoRegex, ".")
        private var mLength: Int = 0

        override fun next(): Interval {
            val part = Interval(mStr, mLength)
            mLength += part.value.length
            return part
        }

        override fun hasNext() = mLength < mStr.length

    }
}
